File: ts/src/00-oops/02-encapsulation/bank-account.ts
/**
 * Encapsulation: Bundling data and methods together
 *
 * Controls access to internal state and ensures data integrity
 * through controlled public interfaces.
 */

class BankAccount {
  private accountNumber: string;
  private balance: number;
  private transactionHistory: Array<{ type: string; amount: number; date: Date }> = [];
  private readonly minBalance: number = 100;

  constructor(accountNumber: string, initialBalance: number) {
    this.accountNumber = accountNumber;
    this.balance = initialBalance;
    this.recordTransaction('deposit', initialBalance);
  }

  deposit(amount: number): boolean {
    if (amount <= 0) {
      console.log('❌ Deposit amount must be positive');
      return false;
    }

    this.balance += amount;
    this.recordTransaction('deposit', amount);
    console.log(`✅ Deposited $${amount}. New balance: $${this.balance}`);
    return true;
  }

  withdraw(amount: number): boolean {
    if (amount <= 0) {
      console.log('❌ Withdrawal amount must be positive');
      return false;
    }

    if (this.balance - amount < this.minBalance) {
      console.log(`❌ Cannot withdraw $${amount}. Minimum balance: $${this.minBalance}`);
      return false;
    }

    this.balance -= amount;
    this.recordTransaction('withdrawal', amount);
    console.log(`✅ Withdrew $${amount}. New balance: $${this.balance}`);
    return true;
  }

  getBalance(): number {
    return this.balance;
  }

  getAccountNumber(): string {
    return this.accountNumber;
  }

  getTransactionHistory(): Array<{ type: string; amount: number; date: Date }> {
    return [...this.transactionHistory];
  }

  private recordTransaction(type: string, amount: number): void {
    this.transactionHistory.push({
      type,
      amount,
      date: new Date(),
    });
  }
}

// Usage
const account = new BankAccount('ACC-12345', 1000);
console.log(`Initial balance: $${account.getBalance()}`);

account.deposit(500);
account.withdraw(300);
account.withdraw(700); // Should fail - exceeds minimum

// Cannot access private fields directly (TypeScript error at compile time)
// account.balance = -1000; // Error!

console.log(`\nFinal balance: $${account.getBalance()}`);
console.log(`\nTransaction history:`);
account.getTransactionHistory().forEach((t) => {
  console.log(`  ${t.type.toUpperCase()}: $${t.amount} - ${t.date.toLocaleTimeString()}`);
});

File: java/src/main/java/com/lld/oops/encapsulation/BankAccount.java
package com.lld.oops.encapsulation;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

/**
 * Encapsulation: Bundling data and methods together with controlled access.
 */
public class BankAccount {
    private String accountNumber;
    private double balance;
    private List<Transaction> transactionHistory;
    private static final double MIN_BALANCE = 100.0;

    public BankAccount(String accountNumber, double initialBalance) {
        this.accountNumber = accountNumber;
        this.balance = initialBalance;
        this.transactionHistory = new ArrayList<>();
        recordTransaction("deposit", initialBalance);
    }

    public boolean deposit(double amount) {
        if (amount <= 0) {
            System.out.println("❌ Deposit amount must be positive");
            return false;
        }

        balance += amount;
        recordTransaction("deposit", amount);
        System.out.println("✅ Deposited $" + amount + ". New balance: $" + balance);
        return true;
    }

    public boolean withdraw(double amount) {
        if (amount <= 0) {
            System.out.println("❌ Withdrawal amount must be positive");
            return false;
        }

        if (balance - amount < MIN_BALANCE) {
            System.out.println("❌ Cannot withdraw $" + amount + ". Minimum balance: $" + MIN_BALANCE);
            return false;
        }

        balance -= amount;
        recordTransaction("withdrawal", amount);
        System.out.println("✅ Withdrew $" + amount + ". New balance: $" + balance);
        return true;
    }

    public double getBalance() {
        return balance;
    }

    public String getAccountNumber() {
        return accountNumber;
    }

    public List<Transaction> getTransactionHistory() {
        return new ArrayList<>(transactionHistory);  // Return copy to prevent modification
    }

    private void recordTransaction(String type, double amount) {
        transactionHistory.add(new Transaction(type, amount, LocalDateTime.now()));
    }

    // Inner class for transaction
    public static class Transaction {
        public final String type;
        public final double amount;
        public final LocalDateTime timestamp;

        public Transaction(String type, double amount, LocalDateTime timestamp) {
            this.type = type;
            this.amount = amount;
            this.timestamp = timestamp;
        }

        @Override
        public String toString() {
            return type.toUpperCase() + ": $" + amount + " - " + timestamp.toLocalTime();
        }
    }
}
